feat(idempotency): Allow durable functions to replay

ConnorKirk opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** fixes #7762 

## Summary
This PR adds support for durable functions in the Idempotency utility. 

1. Introduce the `is_replay` variable. We use this to determine if a function is being replayed.
2. Correctly register the lambda context when a durable context is passed.

### Changes

> Please provide a summary of what's being changed
* Add `is_replay` parameter to the `handle` method
* If there's an existing idempotency record, but the function is a replay, then allow the function to continue
* Add a stubbed `DurableContext` class
* Add tests

### User experience

Idempotency utility now allows durable functions to execute after suspension. 

## Notes

This leaves one issue remaining in the idempotency utility when used with durable functions.

We use `context.getTimeInMillis()` to calculate the expiry time of the `INPROGRESS` status. Duplicate payloads received after that expiry time will be allowed. In a durable execution, the length of the workflow can be up to 1 year, across multiple lambda invocations. This is longer than the maximum 15 minutes that an `INPROGRESS` status can currently have. Idempotency records for long running workflows will not be protected from duplicate invokes beyond the first 15 minutes of their life.

In future, we could
1. Push for the inclusion of the `DurableContext.getExecutionTimeoutInMillis()` method. Once this is added, we can calculate the workflow timeout in the same manner we do for a single invocation.
2. Allow the user to configure a timeout. We'd direct users to set this to match their durable execution timeout. I'm reluctant to add a new configuation option to the Handler/Config, especially if it might be redundant in the medium term (see #1)
3. Retrieve the durable execution configuration within the function and set the timeout. This appears simpler, but requires an additional IAM permission, and might be surprising for a user.


<!-------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/aws-powertools/powertools-lambda-python/blob/main/CONTRIBUTING.md#sending-a-pull-request
- Check that there isn't already a PR that addresses the same issue. If you find a duplicate, please leave a comment under the existing PR so we can discuss how to move forward
- Check that the change meets the project's tenets https://docs.powertools.aws.dev/lambda/python/latest/#tenets
- Add a PR title that follows the conventional commit semantics - https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml
- If relevant, add tests that prove that the change is effective and works
------->

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

ConnorKirk committed:
Support replays in Idempotency utility

Durable Functions introduces the concept of function replays.
Previously, the idempotency utility would throw an
"IdempotencyItemAlreadyExistsError", as the replay has the same payload
of the initial execution. It appears as a duplicate, so is rejected.
Now, a replay is allowed

ConnorKirk committed:
Fix logic in idempotent function decorator

ConnorKirk committed:
Update to record `is_replay`

ConnorKirk committed:
Fix typing

ConnorKirk committed:
Add tests

leandrodamascena commented on aws_lambda_powertools/utilities/typing/lambda_context.py:
```diff
@@ -93,3 +93,16 @@ def tenant_id(self) -> str | None:
+
+    @property
+    def state(self) -> object:
+        return self._state
```
Changed from a concrete class to a Protocol with @runtime_checkable. This avoids type conflicts with the real `DurableContext` from the Durable SDK - users can use the SDK's class directly without import collisions and we avoid creating a type only for this.

```suggestion

# Protocol enables duck typing without competing with the real DurableContext from the SDK
@runtime_checkable
class DurableContextProtocol(Protocol):
    """Protocol for durable execution context compatibility."""

    @property
    def lambda_context(self) -> LambdaContext:
        """The underlying Lambda context."""
        ...

    @property
    def state(self) -> object:
        """The durable execution state."""
        ...
```

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/idempotency.py:
```diff
@@ -91,7 +91,17 @@ def handler(event, context):
+    else:
+        # Standard LambdaContext
+        config.register_lambda_context(context)
+        is_replay = False
```
I think if we change `hasattr(context, "state")` to `hasattr(context, "state") and hasattr(context, "lambda_context")`. We have more explicit check that ensures we only treat it as a durable context when both required attributes exist.

```suggestion
    # Check for durable context using duck typing
    if hasattr(context, "state") and hasattr(context, "lambda_context"):
        durable_context = cast("DurableContextProtocol", context)
        config.register_lambda_context(durable_context.lambda_context)
        # Note: state.operations is accessed via duck typing at runtime
        is_replay = len(durable_context.state.operations) > 1  # type: ignore[attr-defined]
    else:
        # Standard LambdaContext
        config.register_lambda_context(context)
        is_replay = False
```

leandrodamascena reviewed (commented):
Hey @ConnorKirk! Overall the PR looks good! I'd suggest considering these changes to use a Protocol instead of a concrete class - this avoids type conflicts with the real DurableContext from the Durable SDK and enables proper duck typing. I haven't fully tested with mypy, but we shouldn't have any issues.

If you think that make sense, you need to change the class name in all the other files.

Thanks a lot for working on this.

ConnorKirk committed:
Use Protocol for type checking

ConnorKirk committed:
Add more robust check to idempotency

ConnorKirk commented:
Thank you for the review @leandrodamascena . I've incorporated your feedback about using protocols. I have also rebased the branch.
Finally - I slightly tightened the check when we determine whether a call is a replay.

ConnorKirk committed:
Remove unused properties.

ConnorKirk committed:
Exclude DurableContextProtocol from coverage

leandrodamascena committed:
Merge branch 'develop' into feat/df-idempotency

leandrodamascena committed:
Merge branch 'develop' into feat/df-idempotency

leandrodamascena reviewed (approved):
thanks a lot for working on this @ConnorKirk! APPROVED
